src/packet/bwe/time.rs

```diff
@@ -0,0 +1,739 @@
+use std::cmp::Ordering;
+use std::ops::{Add, AddAssign, Div, Neg as _, Sub, SubAssign};
+use std::time::{Duration, Instant};
+
+use crate::bwe::Bitrate;
+use crate::rtp_::DataSize;
+
+/// Wrapper for [`Instant`] that provides additional time points in the past or future.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum SuperInstant {
+    /// A time in the past that already happened.
+    DistantPast,
+
+    /// An exact instant.
+    Exact(Instant),
+
+    /// A time in the future that will never happen.
+    DistantFuture,
+}
+
+/// Wrapper for [`Duration`] that can be negative and provides a duration to a
+/// distant future or past.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum TimeDelta {
+    /// Time delta to some event in distant past that already happened.
+    MinusInf,
+
+    /// An exact negative duration.
+    Negative(Duration),
+
+    /// An exact positive duration.
+    Positive(Duration),
+
+    /// Time delta to some event in distant future that will never happen.
+    PlusInf,
+}
+
+impl TimeDelta {
+    pub(super) const ZERO: Self = Self::Positive(Duration::ZERO);
+
+    /// Returns the number of seconds contained by this [`TimeDelta`] as `f64`.
+    pub fn as_secs_f64(&self) -> f64 {
+        match self {
+            Self::Negative(d) => d.as_secs_f64().neg(),
+            Self::Positive(d) => d.as_secs_f64(),
+            Self::PlusInf => f64::INFINITY,
+            Self::MinusInf => f64::NEG_INFINITY,
+        }
+    }
+}
+
+#[cfg(test)]
+impl TimeDelta {
+    /// Creates a [`TimeDelta`] from seconds.
+    pub const fn from_secs(secs: i64) -> TimeDelta {
+        if secs >= 0 {
+            Self::Positive(Duration::from_secs(secs as u64))
+        } else {
+            Self::Negative(Duration::from_secs(-secs as u64))
+        }
+    }
+
+    /// Creates a [`TimeDelta`] from milliseconds.
+    pub const fn from_millis(millis: i64) -> Self {
+        if millis >= 0 {
+            Self::Positive(Duration::from_millis(millis as u64))
+        } else {
+            Self::Negative(Duration::from_millis(-millis as u64))
+        }
+    }
+}
+
+impl SuperInstant {
+    /// Indicates whether this [`SuperInstant`] is [`SuperInstant::Exact`].
+    pub const fn is_finite(&self) -> bool {
+        matches!(self, Self::Exact(_))
+    }
+
+    /// Indicates whether this [`SuperInstant`] is [`SuperInstant::DistantPast`]
+    /// or [`SuperInstant::DistantFuture`].
+    pub const fn is_not_finite(&self) -> bool {
+        !self.is_finite()
+    }
```

Comment:
I think the word "finite" here is introducing another name for something that is already called "exact" in the enum.

I suggest we stick with one of them. My preference is `is_exact` (do we need `is_not_exact`?)